### ISSUE TITLE
bug(npu): 35B-A3B decode returns token 0 every step — prefill OK, decode argmax broken

### ISSUE THREAD

**Issue by bong-water-water-bong** [1bit-MONSTER/1bit-MONSTER#1830]
## Summary

Running the official Qwen3.6-35B-A3B-NPU2 model (Q4_K_S, `model.q4nx`, 23.2 GB) through `npu_engine_universal` on Strix Halo: prefill works (793 ms/tok, 6 tokens), but **every decode step returns token id 0** (argmax always 0). The `[N] batch=1 toks: 0` lines show the sampled id is 0 for all 8+ decode steps, even with `NPU_GREEDY=1` (deterministic argmax). The model loads, all 40 layers init, xclbins load, prefill logits path works — only the decode argmax is broken.

## Repro

```sh
# tokenize a prompt
./build/engine/npu/npu_engine_universal ~/.config/flm/models/Qwen3.6-35B-A3B-NPU2/model.q4nx 8 /tmp/toks.txt
```

Observed (both `NPU_MOE=0` and `NPU_MOE=1`, with and without `NPU_GREEDY=1`):

```
=== Prefill 6 ===
Prefill: 4761ms (793 ms/tok)
=== M=8 Batch Decode (8 tokens) ===
  [1] batch=1 toks: 0  1025ms
  [2] batch=1 toks: 0  1011ms
  ...
=== 920.5 ms/tok (1 tok/s) | boot=3ms batches=7 tokens=8 ===
```

`boot=0` on the first step too — the argmax id is 0 from step 0.

## Hardware

- AMD Ryzen AI MAX+ 395 (Strix Halo), 128 GB unified
- Ubuntu 26.04, kernel 7.0.0-30-generic
- NPU: XDNA 2, `RyzenAI-npu5`, firmware 1.1.2.65, in-tree amdxdna driver (7.0.0-30)

## Notes

- Model weights: `FastFlowLM/Qwen3.6-35B-A3B-NPU2` `model.q4nx` (23235412536 bytes)
- xclbins: `final_i8_{QKV,O,GU,D}_qwen3_6_35b_a3b.xclbin` + insts (present in `engine/npu/xclbins/`)
- Config detected: `qwen3_6_35b_a3b`, H=2048 NC=40 NH=32 NKV=16 HD=128 IM=512 NV=248320, MoE experts=256 top_k=8 shared=1 gdn=1
- This is the same class of "35B produces garbage" regression noted in `docs/journey.md` (amdxdna 0.16.0 / driver regression history) — the current in-tree driver is the known-good 0.7.0-equivalent, but decode still emits token 0.
- Likely areas: the M=8 batch decode path (`h_b` / `top_ids`), the GatedDeltaNet decode attention, or the MoE expert routing in decode (works at prefill, breaks at decode).

**Comment by bong-water-water-bong** [comment]
Cross-link: the active int4 weight-DMA workstream on [#1769](https://github.com/1bit-MONSTER/1bit-MONSTER/issues/1769) (fused int4 GU→SiLU→D decode on `npu_engine_universal`) is building on the same NPU engine path as this bug. Coordination note posted there — this decode issue may share a root cause with the int4 weight-path work (zaya1-8b kernel round). The int4 pack/dequant primitives (`i4_pack.h`, `q4nx_raw.h`, on-chip dequant) landed in #1793/#1820 are exactly the machinery a fix here would use.

**Comment by bong-water-water-bong** [comment]
**Resolved — stale build artifact. Fresh full rebuild produces a correct token stream.**

Diagnostic session (2026-08-26) on strixhalo, current source (worktree `c22c1b7f`):

1. Reproduced with the stale `build/engine/npu/npu_engine_universal` (Aug 24 CMake build): boot top-5 = `0 1 2 3 4` → all logits identical → argmax 0 every step, exactly as reported.
2. Ruled out via diagnostics + `NPU_DUMP_HIDDEN`:
   - Prefill final hidden (L39, last pos): max|h| = 6.39, real values — prefill is healthy
   - `h_data` at boot: nonzero (`1.18 0.48 2.43 ...`)
   - `fin_v` (final norm weights): nonzero (`2.73 2.77 2.83 ...`)
   - `lm_head`: loaded 31040×2048 from JSON (untied, correct)
3. **Fresh full rebuild** (`engine/npu/build_npu.sh` — recompiles `npu_engine_universal.cpp` + `dequant_q4nx.o` + `gemm_npu_instructions.o` + `zaya_decode.o`):

```
=== Prefill 6 ===
Prefill: 7959ms (1327 ms/tok)
=== M=8 Batch Decode (8 tokens) ===
  [0] boot=6675 (11ms)
  [1] batch=1 toks: 6675   1665ms
  [2] batch=1 toks: 28910  1692ms
  [3] batch=1 toks: 23481  1799ms
  [4] batch=1 toks: 1333   1673ms
  ...
```

Nonzero, varying tokens — argmax/sampling works. The token-0 signature was produced only by the stale Aug 24 binary (the all-equal logits came from outdated dequant/instructions objects).

**Recommendation**: rebuild all variants via `build_npu.sh` (or the CMake path) and re-test — this issue can be closed as fixed-by-rebuild. Left `NPU_DBG`-gated boot diagnostics in the source (`BOOT h_data` / `BOOT fin_v` / `BOOT lg`) for future decode debugging; revert if unwanted.

**Comment by bong-water-water-bong** [comment]
## Related finding (2026-09-03): same token-0 symptom family isolated

This session isolated the 35B 'token 0 every step' symptom in the FastFlowLM RUNTIME path (run_qwen3_6_moe / qwen3_6_moe_npu): prefill + forward return deterministic garbage whose argmax = 0 ('!'), and the layer forward is all-NaN with CORRECT weights + a WORKING lm_head (see #2066 — NaN in both text-only and full-vision configs). Token 0 = '!' is the same degenerate argmax seen here.

If this issue's decode-token-0 was via npu_engine_universal (the 1bit engine's own kernels, not the FastFlowLM lib), it may be a SEPARATE instance of the same underlying 35B format/kernel problem — worth re-testing against #2066's discriminator (does prefill produce REAL logits or NaN-family garbage?). Reopening consideration.
